core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -110,16 +114,60 @@ static double getSourceTablesSizeInBytes(PlanNode node, Lookup lookup, StatsProv
                 .whereIsInstanceOfAny(JoinNode.class, UnnestNode.class)
                 .matches();
         if (hasExpandingNodes) {
-            return Double.NaN;
+            return NaN;
         }
 
         List<PlanNode> sourceNodes = PlanNodeSearcher.searchFrom(node, lookup)
                 .whereIsInstanceOfAny(TableScanNode.class, ValuesNode.class)
                 .findAll();
 
-        return sourceNodes.stream()
+        double sourceTablesSizeInBytes = sourceNodes.stream()
                 .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputSizeInBytes(sourceNode.getOutputSymbols(), typeProvider))
                 .sum();
+
+        double bestKnownRowCount = getBestKnownRowCount(node, lookup, statsProvider);
+        double sourceTablesRowCount = sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputRowCount())
+                .sum();
+
+        if (bestKnownRowCount < sourceTablesRowCount) {
+            return (bestKnownRowCount / sourceTablesRowCount) * sourceTablesSizeInBytes;
+        }
+        return sourceTablesSizeInBytes;
+    }
+
+    @VisibleForTesting
+    static double getBestKnownRowCount(PlanNode node, Lookup lookup, StatsProvider statsProvider)
```

Comment:
why `getBestKnownRowCount` is better than what we had already?

Reply:
I was trying to make the comparisons of size between the two join sides as accurate as possible by taking into account the impact of size reduction from Limit, Filter, Aggregation etc.